### PR TITLE
Coloring  URL in log file

### LINKED ISSUE #153: Show URLs as Hyperlinks in colorized logs
It would be easier to recognize URLs if the text color were blue and underlined.
In addition, it would be better if the URL could be expanded in the browser when clicked.

related to #143.

### PR DESCRIPTION
**Pull request by sahara-ooga**
Add a special coloring in colorized log file for URLs, similar to `nil`, `Bool` or so.

This PR resolves #153.

This PR enhances #143.

## as is 

as-is

## to be

to-be

URL Strings are blue and underlined.

### Tests

There are currently no test cases for coloring in other objects (`Optional.nil`, `Bool`, `String` and Number types) that can be specially colored. 

Therefore, it may be necessary to remove the URL test cases included in this PR or add similar test cases for other types of coloring.

### Terminal ANSI Coloring 

[terminal - Including a shell escape character in a Swift string - Stack Overflow](https://stackoverflow.com/questions/48611540/including-a-shell-escape-character-in-a-swift-string)

[ANSI escape code - Wikipedia](https://en.wikipedia.org/wiki/ANSI_escape_code)

### `XCTUnwrap`

[XCTUnwrap(_:_:file:line:) | Apple Developer Documentation](https://developer.apple.com/documentation/xctest/3380195-xctunwrap)

[Xcode 11でのテスト周りの新機能を紹介します！ - DeNA Testing Blog](https://swet.dena.com/entry/2019/10/23/080000) (In Japanese only)

[How to check and unwrap optionals in tests using XCTUnwrap()](https://www.hackingwithswift.com/example-code/testing/how-to-check-and-unwrap-optionals-in-tests-using-xctunwrap)

### CODE REVIEW

File: Sources/Core/PrettyDescriber.swift [hunk 1]
```diff
@@ -142,9 +142,9 @@ struct PrettyDescriber {
 
         case let url as URL:
             if debug {
-                return theme.typeName("URL") + #"("\#(url.absoluteString)")"#
+                return theme.typeName("URL") + #"("\#(theme.urlLiteral(url.absoluteString))")"#
             } else {
-                return url.absoluteString
+                return theme.urlLiteral(url.absoluteString)
```
Comment (sahara-ooga) [inline Sources/Core/PrettyDescriber.swift]: Only `URL.absoluteString` is subject to dedicated coloring for URL.

File: Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    ///   - url: a url string which is the target to generate strings
+    ///   - expect: the string that is expected to be generated
+    ///   - debug: whether generated string is a detailed version for debugging or not
+    /// - Throws: when the URL could not be generated from the string given in the `url` argument
+    private func _test(
+        _ describer: PrettyDescriber,
+        url string: String,
+        expect: String,
+        debug: Bool) throws {
+        let url = try XCTUnwrap(URL(string: string))
```
Comment (sahara-ooga) [inline Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift]: In this case, I use `XCTUnwrap` function for assert that generated URL is not `nil`.

File: Sources/Core/Color/ColorTheme.swift [hunk 3]
```diff
@@ -14,33 +14,38 @@ public struct ColorTheme {
...
+        urlLiteral: { $0 }
     )
     public static let `default` = ColorTheme(
         typeName: { $0.yellow() },
         nilLiteral: { $0.red() },
         boolLiteral: { $0.blue() },
         stringLiteral: { $0.cyan() },
-        numberLiteral: { $0.green() }
+        numberLiteral: { $0.green() },
+        urlLiteral: { $0.blue().underline() }
```
Comment (YusukeHosonuma) [inline Sources/Core/Color/ColorTheme.swift]: URL is not **Literal** in Swift.
Therefore I suggest to change name to simply `url`.

File: Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift (hunk 2 above)
Comment (YusukeHosonuma) [reply Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift]: This String of URL is given as pre-conditions of test at [L25](https://github.com/YusukeHosonuma/SwiftPrettyPrint/pull/155/files#diff-0bbb03358059b462570b3215cf996bb5b8f4165949fdaec4021290f9719a7e10R25).
So **force-unwrap** is always succeeded in tests.

Therefore I think to use **force-unwrap** is better than it.

```swift
let url = URL(string: string)!
```

or change type of variable `exempleURL` in L25 to `URL`.

```swift
let exampleURL = URL(string: "https://example.com")!
```

(I prefer the latter)

File: Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    override func tearDownWithError() throws {}
+
+    
+    /// URL coloring
+    /// - Throws: when the URL could not be generated from the string given in the `url` argument
+    func testURLColoring_non_debug() throws {
+        try _test(
+            describerWithDefaultColorTheme,
+            url: exampleURL,
+            expect: "\u{1B}[4m\u{1B}[34m\(exampleURL)\u{1B}[0m\u{1B}[4m\u{1B}[0m",
```
Comment (YusukeHosonuma) [inline Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift]: The validity of ANSI-Escape string is guaranteed by [ColorlizeSwift](https://github.com/mtynior/ColorizeSwift).
Also this test-case is test `.default` theme only.

I suggest to create custom `ColorTheme` for tests.

For example:

```swift
let theme = {
    var t = ColorTheme.plain
    t.urlLiteral = { s -> "<URL>\(s)</URL>" }
    return t
}()
...
try _test(
    describerWithTestTheme,
    url: exampleURL,
    expect: "<URL>\(exampleURL)</URL>",
    debug: false
)
```

By doing this, test any `ColorTheme` works correctly (and test codes are clearly).

**Review by YusukeHosonuma — CHANGES_REQUESTED** [review]
Good job!

Note:
I'd like to organize `ColorTheme` tests soon.

**Comment by sahara-ooga** [comment]
The compile errors in the Linux environment are resolved by PR #156.

File: Sources/Core/Color/ColorTheme.swift (hunk 3 above)
Comment (sahara-ooga) [reply Sources/Core/Color/ColorTheme.swift]: Indeed!

I fixed in 67891f2.

File: Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift (hunk 2 above)
Comment (sahara-ooga) [reply Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift]: This is reflected in 53b3077. Thanks!

File: Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift (hunk 4 above)
Comment (sahara-ooga) [reply Tests/SwiftPrettyPrintTests/Core/URLColoringWithDefaultColorThemeTests.swift]: resolve this in 5c229cd0ecc3c9a3887d971ce89e8b193fe2f8eb

File: Tests/SwiftPrettyPrintTests/Core/URLColoringTests.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    override func tearDown() {}
+    
+    ///
+    /// URL Coloring with Custom Color Theme
+    ///
+    func testWithCustomColorTheme() {
+        let theme: ColorTheme = {
+            var t = ColorTheme.plain
+            t.url = { #"<a href="\#($0)">\#($0)</a>"# }
```
Comment (YusukeHosonuma) [inline Tests/SwiftPrettyPrintTests/Core/URLColoringTests.swift:27]: I feel `a` tag with `href` attribute are redundant little for this test.
I'll consider it at organize test of `ColorTheme`.

File: Sources/Core/Log/LogOutputStream.swift [hunk 6]
```diff
@@ -35,7 +35,7 @@
                 guard let bytes = ptr.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                     return
                 }
-                output.write(bytes, maxLength: data.count)
+                _ = output.write(bytes, maxLength: data.count)
```
Comment (YusukeHosonuma) [inline Sources/Core/Log/LogOutputStream.swift:38]: 👍

**Review by YusukeHosonuma — APPROVED** [review]
👍

File: Tests/SwiftPrettyPrintTests/Core/URLColoringTests.swift (hunk 5 above)
Comment (sahara-ooga) [reply Tests/SwiftPrettyPrintTests/Core/URLColoringTests.swift:27]: I understand.

**Pull request merged**